### DISCUSSION TITLE
Key exchange fails when trying to build an IPsec-VPN with StorngSwan and Yamaha vRX

### DISCUSSION THREAD

**Discussion by Rentaro-K** [strongswan/strongswan#1359] [Q&A]
■Summary
Attempts to build an IPsec-VPN with Yamaha vRX in Strongswan have failed.
Where can I check the logs and configuration values to get more information to solve the problem?

 There is a virtual router vRX produced by Yamaha that runs on AWS.

My objective is to build an IPsec-VPN between vRX and StrongSwan.

As a preliminary step, I have tried to build an IPsec-VPN between StrongSwan and another StrongSwan, which was successful.
Next, I changed only the right and rightsubnet values of StrongSwan and tried to connect to the kitted vRX.

The following is the content of StrongSwan's ipsec.conf

> config setup
> 
> conn %default
> 
> conn strongswan
>          authby=psk
>          dpdaction=restart
>          dpddelay=60s
>          dpdtimeout=240s
>          auto=start
>          leftsubnet=10.11.0.0/20
>          right=AAA.AAA.AAA.AAA
>          rightsubnet=192.168.1.0/24
>          keyexchange=ikev1
>          ikelifetime=28800s
>          ike=aes256-sha1-modp2048
>          aggressive=no
>          type=tunnel
>          lifetime=3600s
>          esp=aes256-sha1-modp2048
>          leftid=%any
>          rightid=%any
> 

        

I tried pinging the vRX private IP address from the StrongSwan server with no success.
Next, I checked the status and it appears that the key exchange is failing.
AAA.AAA.AAA.AAA is the global IP address of vRX.

> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[IKE] IKE_SA strongswan[1] established between 10.11.7.226[10.11.7.226]...AAA.AAA.AAA.AAA[192.168.2.254]
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[IKE] IKE_SA strongswan[1] established between 10.11.7.226[10.11.7.226]...AAA.AAA.AAA.AAA[192.168.2.254]
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[IKE] scheduling reauthentication in 28007s
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[IKE] maximum IKE_SA lifetime 28547s
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[IKE] DPD not supported by peer, disabled
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[ENC] generating QUICK_MODE request 1622784197 [ HASH SA No KE ID ID ]
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 11[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (444 bytes)
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 12[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (92 bytes)
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 12[ENC] parsed INFORMATIONAL_V1 request 3911464810 [ HASH N(NO_PROP) ]
> Oct 21 07:11:11 ip-10-11-7-226 charon[7721]: 12[IKE] received NO_PROPOSAL_CHOSEN error notify
> Oct 21 07:11:17 ip-10-11-7-226 charon[7721]: 06[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (188 bytes)
> Oct 21 07:11:17 ip-10-11-7-226 charon[7721]: 06[ENC] parsed QUICK_MODE request 3950772171 [ HASH SA No ]
> Oct 21 07:11:17 ip-10-11-7-226 charon[7721]: 06[IKE] no matching CHILD_SA config found for AAA.AAA.AAA.AAA/32 === 10.11.7.226/32
> Oct 21 07:11:17 ip-10-11-7-226 charon[7721]: 06[ENC] generating INFORMATIONAL_V1 request 4114941704 [ HASH N(INVAL_ID) ]
> Oct 21 07:11:17 ip-10-11-7-226 charon[7721]: 06[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (76 bytes)
> Oct 21 07:11:22 ip-10-11-7-226 charon[7721]: 08[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (188 bytes)
> Oct 21 07:11:22 ip-10-11-7-226 charon[7721]: 08[IKE] received retransmit of request with ID 3950772171, but no response to retransmit

I would like to know which specific settings are causing the key change to fail.
Could you give us a detailed view of the logs and what settings to check?

The log on the vRX side is also included below for reference.
BBB.BBB.BBB.BBB is the global IP address of the StrongSwan server.

> 2022/10/21 16:11:10: [IKE] respond ISAKMP phase to BBB.BBB.BBB.BBB
> 2022/10/21 16:11:10: [IKE] add ISAKMP context [199] 09660e6cde2c5aa1 00000000
> 2022/10/21 16:11:10: [IKE] invalid ISAKMP proposal
> 2022/10/21 16:11:10: [IKE] ISAKMP SA attribute (group description) 31
> 2022/10/21 16:11:10: [IKE] invalid ISAKMP proposal
> 2022/10/21 16:11:10: [IKE] ISAKMP SA attribute (group description) 31
> 2022/10/21 16:11:10: [IKE] add ISAKMP SA[2] (gateway[1])
> 2022/10/21 16:11:10: [IKE] ISAKMP socket[1] is active
> 2022/10/21 16:11:10: [IKE] respond IPsec phase to BBB.BBB.BBB.BBB
> 2022/10/21 16:11:10: [IKE] add IPsec context [200] 09660e6cde2c5aa1 60b9b8c5
> 2022/10/21 16:11:10: [IKE] no proposal chosen []
> 2022/10/21 16:11:10: [IKE] initiate info exchange (notification)
> 2022/10/21 16:11:10: [IKE] inactivate context [200] 09660e6cde2c5aa1 60b9b8c5
> 2022/10/21 16:11:10: [IKE] delete IPsec context [200] 09660e6cde2c5aa1 60b9b8c5
> 2022/10/21 16:11:14: [IKE] SA[1] change state to DEAD
> 2022/10/21 16:11:14: [IKE] delete IPsec context [196] 6d7fd6f94348a8ef 27fe79b6
> 2022/10/21 16:11:14: [IKE] delete ISAKMP context [193] 6d7fd6f94348a8ef 00000000
> 2022/10/21 16:11:14: [IKE] delete SA[1]
> 2022/10/21 16:11:15: [IKE] initiate IPsec phase to BBB.BBB.BBB.BBB
> 2022/10/21 16:11:15: [IKE] add IPsec context [202] 09660e6cde2c5aa1 eb7bffcb
> 2022/10/21 16:11:16: [IKE] SPI 00000000
> 2022/10/21 16:11:16: [IKE] invalid ID information : no message
> 2022/10/21 16:11:21: [IKE] [202] retransmit to BBB.BBB.BBB.BBB (count = 9)
> 2022/10/21 16:11:28: [IKE] [202] retransmit to BBB.BBB.BBB.BBB (count = 8)

For your reference, here is the vRX config.

> console lines infinity
> login timer clear
> vrx user xxx *
> ip route default gateway dhcp lan2
> ip route 10.11.0.0/20 gateway tunnel 1
> ip lan1 address dhcp
> ip lan2 address dhcp
> ip lan2 nat descriptor 1
> tunnel select 1
>  ipsec tunnel 1
>   ipsec sa policy 1 1 esp
>   ipsec ike version 1 1
>   ipsec ike encryption 1 aes256-cbc
>   ipsec ike group 1 modp2048
>   ipsec ike hash 1 sha
>   ipsec ike local address 1 192.168.1.254
>   ipsec ike pre-shared-key 1 *
>   ipsec ike remote address 1 BBB.BBB.BBB.BBB
>  tunnel enable 1
> nat descriptor type 1 masquerade
> nat descriptor masquerade static 1 1 192.168.1.254 udp 500
> nat descriptor masquerade static 1 2 192.168.1.254 udp 4500
> nat descriptor masquerade static 1 3 192.168.1.254 tcp 22
> ipsec use on
> ipsec auto refresh on
> ipsec ike keepalive use 2 on rfc4306
> syslog debug on
> telnetd service off
> dns server dhcp lan2
> sshd service on
> sshd host key generate *

**Comment by tobiasbrunner** [comment]
>  esp=aes256-sha1-modp2048

It could be the DH group (`modp2048`). Does the router expect a DH exchange for phase 2/quick mode? (Might be called perfect-forward secrecy or PFS in its settings.)

Does the Yamaha vRX support IKEv2?

**Reply by Rentaro-K** [reply]
Thanks for the response.

I added a note to the article about the vRX config.

We are still configured to use modp2048 in vRX.
vRX supports IKEv2.
However, due to requests from external parties, there are circumstances where vRX must use IKEv1.

**Reply by Thermi** [reply]
the key exchange in IKE is mandatory, the one for the CHILD_SAs is optional and is called PFS. Verify PFS is on. The configuration you provided indicates it is not explicitely activated. Thus probably disabled.

**Reply by Rentaro-K** [reply]
To Tobias & Thermi


Thanks for the response.

Only explicitly enabling PFS did not solve the problem.
However, explicitly enabling various settings in vRX solved the problem at the same time.

I have not been able to identify which parameters were the cause, but the configuration changes I made are described below.

■StorngSwan side
Add
> pfs=yes
> pfsgroup=modp2048

Deleted
> rekey=yes


■YAMAHA vRX side
Add
> ipsec sa policy 1 1 esp aes256-cbc sha-hmac
> ipsec ike duration ipsec-sa 1 3600
> ipsec ike duration ike-sa 1 28800
> ipsec ike keepalive use 1 on dpd
> ipsec ike local name 1 AAA.AAA.AAA.AAA ipv4-addr
> ipsec ike pre-shared-key 1 text XXXXX
> ipsec ike remote name 1 BBB.BBB.BBB.BBB ipv4-addr

Deleted
> ipsec sa policy 1 1 esp
> ipsec ike pre-shared-key 1 *


You have been very helpful.
Thank you for your advice on this issue.

**Reply by Thermi** [reply]
The settings you added on the strongSwan side are no-ops in that version (unless you downgraded to 4.x). They don't do anything. PFS is configured with a DH group specified as part of the chosen esp proposal.

**Reply by Rentaro-K** [reply]
Am I correct in understanding that this configuration change has not effectively changed the behaviour of StrongSwan?
We understand that the cause of this problem was not in StrongSwan, but in vRX.

**Reply by Thermi** [reply]
Yes, the behavior was not effectively changed. You should review the logs to see what currently happens.

**Reply by Rentaro-K** [reply]
The StrongSwan logs currently displayed are as follows.

The StrongSwan server and the vRX server are successfully pinging each other with a private IP address as the destination.

> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 11[IKE] maximum IKE_SA lifetime 28345s
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 11[ENC] generating QUICK_MODE request 200411346 [ HASH SA No KE ID ID ]
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 11[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (444 bytes)
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (476 bytes)
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[ENC] parsed QUICK_MODE response 200411346 [ HASH SA No KE ID ID ]
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/MODP_2048/NO_EXT_SEQ
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[IKE] CHILD_SA strongswan{1} established with SPIs c6793821_i 1c6b7e84_o and TS 10.11.0.0/20 === 192.168.1.0/24
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[IKE] CHILD_SA strongswan{1} established with SPIs c6793821_i 1c6b7e84_o and TS 10.11.0.0/20 === 192.168.1.0/24
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[ENC] generating QUICK_MODE request 200411346 [ HASH ]
> Oct 25 06:25:07 ip-10-11-7-226 charon[1117]: 12[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (60 bytes)
> Oct 25 06:25:12 ip-10-11-7-226 charon[1117]: 08[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (92 bytes)
> Oct 25 06:25:12 ip-10-11-7-226 charon[1117]: 08[ENC] parsed INFORMATIONAL_V1 request 1002347246 [ HASH N(DPD) ]
> Oct 25 06:25:12 ip-10-11-7-226 charon[1117]: 08[ENC] generating INFORMATIONAL_V1 request 2812862561 [ HASH N(DPD_ACK) ]
> Oct 25 06:25:12 ip-10-11-7-226 charon[1117]: 08[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (92 bytes)
> Oct 25 06:25:22 ip-10-11-7-226 charon[1117]: 05[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (92 bytes)
> Oct 25 06:25:22 ip-10-11-7-226 charon[1117]: 05[ENC] parsed INFORMATIONAL_V1 request 1572657068 [ HASH N(DPD) ]
> Oct 25 06:25:22 ip-10-11-7-226 charon[1117]: 05[ENC] generating INFORMATIONAL_V1 request 3801479101 [ HASH N(DPD_ACK) ]
> Oct 25 06:25:22 ip-10-11-7-226 charon[1117]: 05[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (92 bytes)
> Oct 25 06:39:32 ip-10-11-7-226 charon[1117]: 10[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (92 bytes)
> Oct 25 06:39:32 ip-10-11-7-226 charon[1117]: 10[ENC] parsed INFORMATIONAL_V1 request 1456701024 [ HASH N(DPD) ]
> Oct 25 06:39:32 ip-10-11-7-226 charon[1117]: 10[ENC] generating INFORMATIONAL_V1 request 2744605865 [ HASH N(DPD_ACK) ]
> Oct 25 06:39:32 ip-10-11-7-226 charon[1117]: 10[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (92 bytes)
> Oct 25 06:39:42 ip-10-11-7-226 charon[1117]: 11[NET] received packet: from AAA.AAA.AAA.AAA[500] to 10.11.7.226[500] (92 bytes)
> Oct 25 06:39:42 ip-10-11-7-226 charon[1117]: 11[ENC] parsed INFORMATIONAL_V1 request 2424317444 [ HASH N(DPD) ]
> Oct 25 06:39:42 ip-10-11-7-226 charon[1117]: 11[ENC] generating INFORMATIONAL_V1 request 1418529211 [ HASH N(DPD_ACK) ]
> Oct 25 06:39:42 ip-10-11-7-226 charon[1117]: 11[NET] sending packet: from 10.11.7.226[500] to AAA.AAA.AAA.AAA[500] (92 bytes)
